### PR TITLE
Alternative way of running doctests

### PR DESCRIPTION
**Pull request by newhoggy**
Previously we had to disable `doctests` on `ghc-8.10.x` because `doctest` doesn't work on the new compiler.  Given that support for `ghc-8.10.x` is going away, we need some way of running these `doctests` with the new compiler.

This PR replaces `doctests` with `cabal-docspec` which does the same thing, except differently so it doesn't suffer the same problem.

One major difference is that everything that get tested in doctest comments must be exported and imported.  Some minor changes were necessary to the doctest comments in order for them to pass.

This PR will allow us to delete `ghc-8.6.5` from all our CI.

### CODE REVIEW

**Comment by newhoggy** [comment]
There are some test failures, but I think they are just flaky tests.

File: semantics/executable-spec/src/Control/State/Transition/Extended.hs [hunk 1]
```diff
@@ -53,6 +53,8 @@ module Control.State.Transition.Extended
 
     -- * Random thing
     Threshold (..),
+
+    sfor_,
```
Comment (JaredCorduan) [inline semantics/executable-spec/src/Control/State/Transition/Extended.hs:57]: just out of curiosity, what was this needed for?

Comment (newhoggy) [reply semantics/executable-spec/src/Control/State/Transition/Extended.hs:57]: Unfortunately `cabal-docspec` requires that every identifier used in the doctest is exported.

Comment (newhoggy) [reply semantics/executable-spec/src/Control/State/Transition/Extended.hs:57]: See the end of the file:

```
-- | 'sfor_' is 'straverse_' with its arguments flipped. For a version
-- that doesn't ignore the results see 'Data.Traversable.for'.
--
-- >>> sfor_ ([1..4] :: [Int]) print    <---------------------------- use of this doctest requires `sfor_` to be exported
-- 1
-- 2
-- 3
-- 4
sfor_ :: (Foldable t, Applicative f) => t a -> (a -> f b) -> f ()
{-# INLINE sfor_ #-}
sfor_ = flip straverse_
```

Comment (newhoggy) [reply semantics/executable-spec/src/Control/State/Transition/Extended.hs:57]: Documentation describing this restriction is found here: https://github.com/phadej/cabal-extras/blob/master/cabal-docspec/MANUAL.md

Comment (newhoggy) [reply semantics/executable-spec/src/Control/State/Transition/Extended.hs:57]: > cabal-docspec doesn't interpret library code in ghci, but instead loads precompiled code. The effect is similar as using -fobject-code in GHCi (which is mandatory for packages with FFI, for example). The consequence is that cabal-docspec is unable to evaluate doctest examples in non-exported modules (other-modules), or which use non-exposed symbols.

**Review by nc6 — APPROVED** [review]

**Comment by newhoggy** [comment]
Thanks!

**Pull request merged**